### DISCUSSION TITLE
File Save dialog in ctrlr 5.6.23

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#379] [Q&A] [answered]
Has anyone successfully managed to save a binary file from ctrlr 5.6.23?

I'm trying to save a .json file and ctrlr crashes and takes down the host (VST Host) immediately after the file save dialog in finished (it does do the check for overwrite) and crashes before any of the checks for fileExists etc happen in my LUA code.

This is my code. I indicate the crash position (as per the debugger stepping through this). The autosave method also crashes when I manually invoke it.  The only difference is that we don't raise the file dialog. The path is programatically determined.

```
function saveJupiterButtonMethod(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
	
	-- This is a manual save of the Library. We will use a different method for autosaving, because we don't want to prompt for 
    -- file location. All library files will be saved with a .jp8, but it will be json files.  We don't want to import any json files, as we don't
    -- do checking at the moment. It would almost certainly break stuff.

         if isPanelReady() == false then -- check panel readiness
		    return
	    end

        if value == 0 then -- stops double triggering when button set as momentary.
            return
        end -- if

        local bankToSave = jupiterBank
        local notify = "Jupiter"

        if type(bankToSave) ~= "table" then
            utils.warnWindow("No Bank To Save", notify.." Bank is Empty. It won't be saved.")
            --buttonName:setValue(0,false,false)
            return
        end --if
        savePath = L(panel:getLabelComponent("filePathLabel"):getPropertyString("uiLabelText"))

        local exportJson = encodeBank(bankToSave)

        -- fileToWrite = utils.saveFileWindow("Save Library to .jp8 (json) format", File.getSpecialLocation(File.userDocumentsDirectory), "*.jp8", true)
        if panel:getLabelComponent("filePathLabel"):getPropertyString("uiLabelText") == nil then
            fileToWrite = utils.saveFileWindow("Save "..notify.." to .jp8 (json) format", File.getSpecialLocation(File.userDocumentsDirectory), "*.jp8", true)
        else
        --local tempPath = savePath..saveFileName
        fileToWrite = utils.saveFileWindow("Save "..notify.." to .jp8 (json) format", File(savePath), "*.jp8", true)
        --fileToWrite = File(savePath..saveFileName)
        end --if

-- ****************** This is where it crashes **************************

        if fileToWrite:isValid() == false then
            --buttonName:setValue(0,false,false)
		    return
	    end

	    -- Let's see if the file exists
	    if fileToWrite:existsAsFile() == false then
		
		    -- The file is not there, that's ok, let's try to create it
		    if fileToWrite:create() == false then
			    -- Ooooh we can't create it, we need to fail here
			    utils.warnWindow ("File write", "The destination file does not exist, and i can't create it")
			    return
		    end
	    end
    
        -- If we reached this point, we have a valid file we can try to write to
	
	    -- There are two falgs when writing text
	    -- asUnicode - if should write our text as UNICODE
	    -- writeUnicodeHeaderBytes - if we should add a special unicode header at the beginning of the file
	    -- we set both to false

        -- Update the file path
        local fullPath = fileToWrite:getFullPathName()
        local fileName = fileToWrite:getFileName()
        lastPath = string.gsub(fullPath,fileName,"")
        -- panel:getLabelComponent("filePathLabel"):setText (lastPath)

        if fileToWrite:replaceWithText (exportJson, false, false) == false then
		    utils.warnWindow ("File write", "Failed to write data to file: "..fileToWrite.getFullPathName())
	    end
end
```

**Comment by Youngwhisk** [comment]
Is it possibly a pemissions or lack of Administrator priveledges issue?

**Reply by Godlike-Productions** [reply]
I hadn't thought of that.  I'll give that a try.

**Reply by Youngwhisk** [reply]
Maybe also check to make sure you aren't missing any back slashes in your concatenated path.

**Comment by Godlike-Productions** [comment, marked as answer]
OK, I think I solved this and another issue that turned out to be related.

I had missed cleaning up a bit of code, so a value was being written into the table as nil.  It turns out that trying to display nil in a text box, or saving it to a file crashes things badly (and in 5.6.23, there is no error catcher)  Fixing this nil value fixed the problem, but I probably need to start puting more error catching in my code.

This lead me down a little investigation that led me here

https://www.tutorialspoint.com/lua/lua_error_handling.htm

I think we should use assert and pcall a lot more often in our code.  LUA doesn't seem to do everything in protected mode by default, and VST's don't seem to run in protected mode within VST hosts.  So an error in our code can crash our host.

pcall allows any LUA function to run in protected mode and allows us to throw custom errors, so as a worst case, it should just stop the function from running without crashing your LUA instance and more importantly the host.

I'll also post this in a show and tell thread to make it easier for people to find.
